core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -159,14 +160,16 @@ public int read(byte[] b, int off, int len) throws IOException {
           mPosition += bytesRead;
         }
         retry.reset();
+        lastException = null;
       } catch (UnavailableException | ConnectException | DeadlineExceededException e) {
-        if (!retry.attempt()) {
-          throw e;
-        }
+        lastException = e;
         handleRetryableException(mBlockInStream, e);
         mBlockInStream = null;
       }
     }
+    if (lastException != null) {
```

Comment:
Check `!retry.attempt()` instead (same applies to below)

Reply:
it could be that the last retry succeeded so that `retry.attempt()` is false, but we shouldn't throw.

Reply:
Oh you're right, we should probably audit our code for this (since I think I saw this pattern in other places). I dislike how we need to keep track of an exception object. We could make `retry.attempt()` throw an exception instead of returning false (we could pass this exception in)

Reply:
the exception would still need to be saved at the outer scope though. TBH the original retry mechanism is looking cleaner at this point. What do you think of this pattern?

```java
while (!retry.isFinished()) {
  try {
    // do something
  } catch (Exception e) {
    if (!retry.waitForNext()) {
      throw formatError(e, retry.getAttempts());
    }
    handle(e);
  }
}
```

`isFinished()` doesn't increment the retry count or wait, it just says whether the policy allows more attempts to be made. `waitForNext()` increments the count and waits until the next attempt should happen.

I like the previous approach and the above approach equally.

Reply:
@aaudiber How about ```while (!retry.hasAttempts())```?

Reply:
`retry.hasAttempts()` SGTM

Reply:
Let's handle this in another PR, since it will affect retries in general.